### DISCUSSION TITLE
Generate preview databases on every pull request

### DISCUSSION THREAD

**Discussion by 3x071c** [vercel/community#395] [Idea]
Is there a database provider that automatically creates new database "environments" (isolated sandboxes) for every preview branch that Vercel deploys (like f.e. when deploying with [Railway](https://railway.app))? Right now, most examples I see use a single production database (f.e. a free PostgreSQL one on Heroku), and PRs share a single staging/preview database. This of course doesn't work for production apps, because a database change in one of the preview deployments would affect all other preview deployments, thereby breaking unrelated PRs. I know that [PlanetScale](https://planetscale.com) offers a branching mechanism, but that is manual and doesn't automatically follow the source branches on GitHub or preview deployments on Vercel. I can't believe there's no solution for this out there, what do you guys do?

**Comment by Trake101** [comment]
Have any solutions for this surfaced? We're in the process of moving to Vercel and we can't seem to find any services that do this.

**Reply by 3x071c** [reply]
For [our project](https://github.com/3x071c/lsg-remix), we ended up using [Fly](https://fly.io) instead, and provisioning a separate container and Postgres database for every preview environment using GitHub workflows and the Fly CLI. This requires a lot of script maintenance though, especially because the Fly api is unreliable and workflows occasionally fail for no reason. The DX is clearly inferior to Vercel, but atm it's the only option I see that also offers multi-region local deploys and a lot of flexibility. [Here's a sample script](https://github.com/3x071c/lsg-remix/blob/trunk/.github/workflows/preview.yml). Now that Vercel uses Cloudflare Workers under the hood for edge deploys, you could also try implementing a database system on top of Cloudflare's KV storage. [This is a working implementation of such a system](https://github.com/3x071c/lsg-remix/blob/5cd9b1b9fdce331be50c4d83bb72a69067004f6b/app/models/_handler.ts), but it comes with several drawbacks (inconsistency, no immediate updating -> optimistic UI updates etc. might show stale data after modification)

**Comment by MatteoGauthier** [comment]
@3x071c take a look at https://neon.tech 👍
